### ISSUE TITLE
Support for partial updates using PATCH

### ISSUE THREAD

**Issue by KevinMitchell** [mulesoft-labs/raml-for-jax-rs#257]
When using PATCH for partial updates (rfc7386) null values in the JSON are used to indicate that an existing value should be removed.  However, for this to work our POJOs have to be able to distinguish between properties that are null because they haven't been initialised, and those that have been explicitly set to null.  Serializers should be able to use this information when building the JSON, and deserialisers should be able to set this information to allow the users of the POJOs to distinguish between the "different" nulls.  Obviously not all APIs will want to use the JSON merge patch approach.  But for those that do it would be very convenient if the generated POJOs had support for this approach.  

Perhaps this is already supported and I've just not noticed the setting?

Thanks, Kevin

**Comment by jpbelang** [comment]
Are you thinking along the lines of @JsonInclude ?

**Comment by KevinMitchell** [comment]
With @JsonInclude I thought you could only choose between whether to transmit null values or to skip them.  Whereas to support merge patch it needs to be more refined that that.  You only want to transmit a null property if it's been explicitly set to null.  And similarly when working with a POJO that's been built from a JSON object you need to be able to distinguish between a null property that was built from a JSON property that was set to null, and a null property that was null because the JSON didn't mention it.  Hopefully that makes sense :-)  

In terms of an implementation there are multiple options.  For example, you could have an additional bit set field in an object, containing a bit per field.  The setter for each field would set the bit, you could test whether a field had been explicitly set, and the serialiser and deserialiser for the object would then use the bitset field to decide what to do with null properties on the way out, or JSON properties set to null on the way in.

Kevin

**Comment by jpbelang** [comment]
Yeah.  Thinking like a ramltojaxrs that could allow that, because you would need a custom serializer, correct ?  The default serializer does not do that, so maybe with something like @JsonSerialize ?  (you are correct on @JsonInclude)

So I would need something to skip the "normal" jackson2 annotating and delegating to another plugin.

**Comment by KevinMitchell** [comment]
Yes, you'd need a custom serialiser (and deserialiser) driven by whatever mechanism was used to track the explicit nulls.  Plus modified setters and a method(s) for checking whether something was explicitly set to null.  It's all very tedious to write manually, but ideal for mechanical generation.

**Comment by jpbelang** [comment]
I agree.  A bit more complicated though.  I'll keep you posted.  This WE, I'm closing simpler problems.  I'll see where I get to.

**Comment by KevinMitchell** [comment]
Thanks for considering this request.

> This WE, I'm closing simpler problems. I'll see where I get to.

If you are looking for simpler problems #254 will hopefully involve a one-line fix :-)
